File: src/main/java/io/bindernews/thegrackle/MiscUtil.java
package io.bindernews.thegrackle;

import basemod.AutoAdd;
import basemod.BaseMod;
import basemod.abstracts.CustomPlayer;
import com.badlogic.gdx.Gdx;
import com.evacipated.cardcrawl.mod.stslib.Keyword;
import com.google.gson.Gson;
import com.megacrit.cardcrawl.actions.AbstractGameAction;
import com.megacrit.cardcrawl.core.AbstractCreature;
import com.megacrit.cardcrawl.core.Settings;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.localization.*;
import com.megacrit.cardcrawl.powers.AbstractPower;
import javassist.CtClass;
import lombok.SneakyThrows;
import lombok.val;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.Nullable;

import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * A collection of miscellaneous utility functions that should be relatively re-usable for other
 * Slay the Spire mods. Feel free to copy-paste this file into your own projects, as long as you
 * retain the MIT license.
 *
 * @author bindernews
 */
public class MiscUtil {
    static final Logger log = LogManager.getLogger(MiscUtil.class.getPackage().getName());
    private static final Gson gson = new Gson();

    @Nullable @SneakyThrows
    public static <T> T nullCast(Class<T> clz, Object o) {
        if (clz.isInstance(o)) {
            return clz.cast(o);
        } else {
            return null;
        }
    }


    /**
     * Returns a new array list containing the given elements. This function exists in later versions
     * of Java, but not in Java 8. The original use-case was {@link CustomPlayer#getStartingDeck()}.
     */
    @SafeVarargs
    public static <E> ArrayList<E> arrayListOf(E ...elements) {
        ArrayList<E> lst = new ArrayList<>();
        lst.ensureCapacity(elements.length);
        Collections.addAll(lst, elements);
        return lst;
    }

    /**
     * Add an action to the bottom of the stack, convenience function.
     */
    public static void addToBot(AbstractGameAction a) {
        AbstractDungeon.actionManager.addToBottom(a);
    }

    /**
     * Add an action to the top of the stack, convenient static import.
     */
    public static void addToTop(AbstractGameAction a) {
        AbstractDungeon.actionManager.addToTop(a);
    }

    public static int getPowerAmount(AbstractCreature c, String powerId, int defaultAmount) {
        return Optional.ofNullable(c.getPower(powerId)).map(p -> p.amount).orElse(defaultAmount);
    }


    private static final HashMap<Class<?>, String> stringTypeMap = new HashMap<>();
    static {
        HashMap<Class<?>, String> m = stringTypeMap;
        m.put(BlightStrings.class, "blight");
        m.put(CardStrings.class, "card");
        m.put(CharacterStrings.class, "character");
        m.put(EventStrings.class, "event");
        m.put(MonsterStrings.class, "monster");
        m.put(OrbStrings.class, "orb");
        m.put(PotionStrings.class, "potion");
        m.put(PowerStrings.class, "power");
        m.put(RelicStrings.class, "relic");
        m.put(StanceStrings.class, "stance");
        m.put(UIStrings.class, "ui");
    }

    /**
     * Load all localization strings from the specified language at the path.
     *
     * <p>
     * The language directories must be lowercase, and files are named in the singular and without
     * the 'Strings' suffix. For example {@link CardStrings} would load {@code card.json}, and {@link UIStrings}
     * would load {@code ui.json}. Missing files will be ignored.
     * </p>
     *
     * @param path Resource directory containing languages, with no trailing slash (e.g. mymod/localization)
     * @param language Language code
     */
    public static void loadLocalization(String path, Settings.GameLanguage language) {
        String lang = language.name().toLowerCase();
        for (Class<?> st : stringTypeMap.keySet()) {
            String filePath = path + "/" + lang + "/" + stringTypeMap.get(st) + ".json";
            if (Gdx.files.internal(filePath).exists()) {
                BaseMod.loadCustomStringsFile(st, filePath);
                log.debug("loadLocalization: loaded " + filePath);
            }
        }
    }

    /**
     * Read a JSON file containing an array of Keyword entries
     * @param path The {@code localizations} directory, with no trailing slash
     * @param language Language code
     * @return Array of keywords
     */
    public static Keyword[] readKeywords(String path, Settings.GameLanguage language) {
        val filePath = langPath(path, "keyword.json", language);
        val content = Gdx.files.internal(filePath).readString(StandardCharsets.UTF_8.name());
        return gson.fromJson(content, Keyword[].class);
    }

    /**
     * Calls {@link MiscUtil#readKeywords} and then {@link BaseMod#addKeyword}.
     * @param modId The mod ID to prefix the keywords with
     * @param path The {@code localizations} directory, with no trailing slash
     * @param language Language code
     */
    public static void loadKeywords(String modId, String path, Settings.GameLanguage language) {
        val keywords = MiscUtil.readKeywords(path, language);
        for (val k : keywords) {
            BaseMod.addKeyword(modId, k.PROPER_NAME, k.NAMES, k.DESCRIPTION);
        }
    }

    public static String langPath(String base, String fileName, Settings.GameLanguage language) {
        val lang = language.name().toLowerCase();
        return base + "/" + lang + "/" + fileName;
    }

    public static void registerPower(Class<? extends AbstractPower> clazz) {
        String powerId = null;
        try {
            powerId = (String) clazz.getDeclaredField("POWER_ID").get(null);
        } catch (Exception ignored) {}
        if (powerId == null) {
            try {
                val ctor = clazz.getDeclaredConstructor(AbstractCreature.class, int.class);
                powerId = ctor.newInstance(null, 0).ID;
            } catch (Exception ignore) {}
        }
        if (powerId == null) {
            try {
                val ctor = clazz.getDeclaredConstructor();
                powerId = ctor.newInstance().ID;
            } catch (Exception ignore) {}
        }
        if (powerId == null) {
            throw new RuntimeException("cannot determine power ID for class " + clazz);
        }
        BaseMod.addPower(clazz, powerId);
    }

    @SneakyThrows
    @SuppressWarnings("unchecked")
    public static <T> List<Class<? extends T>> autoFindClasses(AutoAdd aa, Class<T> type) {
        ArrayList<Class<? extends T>> lst = new ArrayList<>();
        for (CtClass c : aa.findClasses(type)) {
            if (c.getAnnotation(AutoAdd.Ignore.class) != null) {
                continue;
            }
            Class<?> c2 = type.getClassLoader().loadClass(c.getName());
            lst.add((Class<? extends T>) c2);
        }
        return lst;
    }

}

File: src/main/kotlin/net/bindernews/grackle/helper/MiscUtil.kt
package net.bindernews.grackle.helper

import basemod.AutoAdd
import basemod.BaseMod
import basemod.ReflectionHacks
import basemod.abstracts.CustomRelic
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.files.FileHandle
import com.evacipated.cardcrawl.mod.stslib.Keyword
import com.google.gson.Gson
import com.megacrit.cardcrawl.actions.AbstractGameAction
import com.megacrit.cardcrawl.core.AbstractCreature
import com.megacrit.cardcrawl.core.Settings.GameLanguage
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.localization.*
import com.megacrit.cardcrawl.powers.AbstractPower
import com.megacrit.cardcrawl.stances.AbstractStance
import net.bindernews.grackle.cards.BaseCard
import net.bindernews.grackle.cards.CardConfig
import net.bindernews.grackle.power.BasePower
import org.apache.logging.log4j.LogManager
import java.util.*

/**
 * A collection of miscellaneous utility functions that should be relatively re-usable for other Slay the Spire mods.
 * <br></br>
 * Feel free to copy-paste this file into your own projects, as long as you retain the MIT license.
 *
 * @author bindernews
 */
object MiscUtil {
    private val log = LogManager.getLogger(MiscUtil::class.java)!!
    private val gson = Gson()
    private val stringTypeMap = HashMap<Class<*>, String>()
    init {
        val m = stringTypeMap
        m[BlightStrings::class.java] = "blight"
        m[CardStrings::class.java] = "card"
        m[CharacterStrings::class.java] = "character"
        m[EventStrings::class.java] = "event"
        m[MonsterStrings::class.java] = "monster"
        m[OrbStrings::class.java] = "orb"
        m[PotionStrings::class.java] = "potion"
        m[PowerStrings::class.java] = "power"
        m[RelicStrings::class.java] = "relic"
        m[StanceStrings::class.java] = "stance"
        m[UIStrings::class.java] = "ui"
    }
    private const val KEYWORD_FILE = "keyword"

    /**
     * Add an action to the bottom of the stack, convenient static import.
     */
    fun addToBot(a: AbstractGameAction?) {
        AbstractDungeon.actionManager.addToBottom(a)
    }

    /**
     * Add an action to the top of the stack, convenient static import.
     */
    fun addToTop(a: AbstractGameAction?) {
        AbstractDungeon.actionManager.addToTop(a)
    }

    fun getPowerAmount(c: AbstractCreature, powerId: String, defaultAmount: Int): Int {
        return c.getPower(powerId)?.amount ?: defaultAmount
    }


    /**
     * Load all localization strings from the specified language at the path.
     *
     * The language directories must be lowercase, and files are named in the singular and without
     * the 'Strings' suffix. For example [CardStrings] would load `card.json`, and [UIStrings]
     * would load `ui.json`. Missing files will be ignored.
     *
     * @param path Resource directory containing languages, with no trailing slash (e.g. "mymod/localization")
     * @param languages Language codes
     */
    fun loadLocalization(path: String, vararg languages: GameLanguage) {
        val seen = HashSet<GameLanguage>()
        for (lang in languages) {
            // In case the same language is given multiple times
            if (!seen.add(lang)) {
                continue
            }
            val langStr = lang.toLowerStr()
            for (st in stringTypeMap.keys) {
                val fileRef = langFile(path, langStr, stringTypeMap[st]!!) ?: continue
                BaseMod.loadCustomStringsFile(st, fileRef.path())
                log.debug("loadLocalization: loaded ${fileRef.path()}")
            }
        }

    }

    /**
     * Reads keywords from 'keyword.json' and then calls [BaseMod.addKeyword] on each.
     * @param modId The mod ID to prefix the keywords with
     * @param path The `localizations` directory, with no trailing slash
     * @param languages Language codes
     */
    fun loadKeywords(modId: String?, path: String, vararg languages: GameLanguage) {
        val seen = HashSet<GameLanguage>()
        for (lang in languages) {
            // In case same language is given multiple times
            if (!seen.add(lang)) {
                continue
            }
            val langStr = lang.toLowerStr()
            val fileRef = langFile(path, langStr, KEYWORD_FILE) ?: continue
            val content = fileRef.readString("UTF-8")
            val keywords = gson.fromJson(content, Array<Keyword>::class.java)
            for (k in keywords) {
                BaseMod.addKeyword(modId, k.PROPER_NAME, k.NAMES, k.DESCRIPTION)
            }
            log.debug("loadKeywords: loaded ${fileRef.path()}")
        }
    }

    private fun GameLanguage.toLowerStr(): String = this.name.lowercase(Locale.getDefault())
    private fun langFile(base: String, langStr: String, fileName: String): FileHandle? {
        return Gdx.files.internal("$base/$langStr/${fileName}.json").takeIf { it.exists() }
    }

    fun getObjectId(clz: Class<*>): String {
        if (AbstractPower::class.java.isAssignableFrom(clz)) {
            @Suppress("UNCHECKED_CAST")
            val powerClazz = clz as Class<out AbstractPower>
            try {
                 return ReflectionHacks.getPrivateStatic(clz, "POWER_ID")
            } catch (_: Exception) {}
            try {
                val ctor = powerClazz.getDeclaredConstructor(AbstractCreature::class.java, Int::class.javaPrimitiveType)
                return ctor.newInstance(null, 0).ID
            } catch (_: Exception) { }
            try {
                val ctor = powerClazz.getDeclaredConstructor()
                return ctor.newInstance().ID
            } catch (_: Exception) { }
        } else if (BaseCard::class.java.isAssignableFrom(clz)) {
            try {
                return ReflectionHacks.getPrivateStatic<CardConfig>(clz, "C").ID
            } catch (_: Exception) {}
        } else if (CustomRelic::class.java.isAssignableFrom(clz)) {
            try {
                return ReflectionHacks.getPrivateStatic(clz, "ID")
            } catch (_: Exception) {}
        } else if (AbstractStance::class.java.isAssignableFrom(clz)) {
            try {
                return ReflectionHacks.getPrivateStatic(clz, "STANCE_ID")
            } catch (_: Exception) {}
        }
        // Fallback returns empty string
        return ""
    }

    fun <T> autoFindClasses(aa: AutoAdd, type: Class<T>): List<Class<out T>> {
        val lst = ArrayList<Class<out T>>()
        for (c in aa.findClasses(type)) {
            if (c.getAnnotation(AutoAdd.Ignore::class.java) != null) {
                continue
            }
            val c2 = type.classLoader.loadClass(c.name)
            @Suppress("UNCHECKED_CAST")
            lst.add(c2 as Class<out T>)
        }
        return lst
    }
}